File: ios/Classes/TiGooglemapsAnnotationProxy.m
/**
 * Ti.GoogleMaps
 * Copyright (c) 2015-Present by Hans Knoechel, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiGooglemapsAnnotationProxy.h"
#import "TiUtils.h"

@implementation TiGooglemapsAnnotationProxy

@synthesize marker = _marker;

- (GMSMarker *)marker
{
    if (!_marker) {
        _marker = [[GMSMarker alloc] init];
        
        CLLocationDegrees latitude = [TiUtils doubleValue:[self valueForKey:@"latitude"]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[self valueForKey:@"longitude"]];
        
        [_marker setPosition:CLLocationCoordinate2DMake(latitude,longitude)];
        [_marker setUserData:@{@"uuid": [[NSUUID UUID] UUIDString]}];
    }
    
    return _marker;
}

- (NSArray *)keySequence
{
    return @[@"latitude", @"longitude"];
}

#pragma mark Public API's

- (TiViewProxy *)infoWindow
{
    return [self valueForUndefinedKey:@"infoWindow"];
}

- (void)setInfoWindow:(id)value
{
    [self replaceValue:value forKey:@"infoWindow" notification:NO];
}

- (void)setLatitude:(id)value
{
    ENSURE_TYPE(value, NSNumber);
    [self replaceValue:value forKey:@"latitude" notification:NO];
}

- (void)setLongitude:(id)value
{
    ENSURE_TYPE(value, NSNumber);
    [self replaceValue:value forKey:@"longitude" notification:NO];
}

- (void)setTitle:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    ENSURE_TYPE(value, NSString);
    
    [[self marker] setTitle:[TiUtils stringValue:value]];
    [self replaceValue:value forKey:@"title" notification:NO];
}

- (void)setSubtitle:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    ENSURE_TYPE(value, NSString);
    
    [[self marker] setSnippet:[TiUtils stringValue:value]];
    [self replaceValue:value forKey:@"subtitle" notification:NO];
}

- (void)setCenterOffset:(id)args
{
    ENSURE_UI_THREAD_1_ARG(args);
    
    [[self marker] setInfoWindowAnchor:[TiUtils pointValue:args]];
    [self replaceValue:args forKey:@"centerOffset" notification:NO];
}

- (void)setGroundOffset:(id)args
{
    ENSURE_UI_THREAD_1_ARG(args);
    
    [[self marker] setGroundAnchor:[TiUtils pointValue:args]];
    [self replaceValue:args forKey:@"groundOffset" notification:NO];
}

- (void)setImage:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    [[self marker] setIcon:[TiUtils image:value proxy:self]];
    [self replaceValue:value forKey:@"image" notification:NO];
}

- (void)setPinColor:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    [[self marker] setIcon:[GMSMarker markerImageWithColor:[[TiUtils colorValue:value] _color]]];
    [self replaceValue:value forKey:@"pinColor" notification:NO];
}

- (void)setTouchEnabled:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    ENSURE_TYPE(value, NSNumber);
    
    [[self marker] setTappable:[TiUtils boolValue:value def:YES]];
    [self replaceValue:value forKey:@"touchEnabled" notification:NO];
}

- (void)setFlat:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    ENSURE_TYPE(value, NSNumber);
    
    [[self marker] setFlat:[TiUtils boolValue:value def:NO]];
    [self replaceValue:value forKey:@"flat" notification:NO];
}

- (void)setDraggable:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    ENSURE_TYPE(value, NSNumber);
    
    [[self marker] setDraggable:[TiUtils boolValue:value def:NO]];
    [self replaceValue:value forKey:@"draggable" notification:NO];
}

- (void)setOpacity:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    ENSURE_TYPE(value, NSNumber);
    
    [[self marker] setOpacity:[TiUtils floatValue:value def:1]];
    [self replaceValue:value forKey:@"opacity" notification:NO];
}

- (void)setAnimationStyle:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    ENSURE_TYPE(value, NSNumber);
    
    [[self marker] setAppearAnimation:[TiUtils intValue:value def:kGMSMarkerAnimationNone]];
    [self replaceValue:value forKey:@"animationStyle" notification:NO];
}

- (void)setRotation:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    [[self marker] setRotation:[TiUtils doubleValue:value def:0]];
}

- (void)setUserData:(id)value
{
    ENSURE_UI_THREAD_1_ARG(value);
    NSMutableDictionary *result = value;
    
    // Hook in internal uuid
    if ([[self marker] userData] != value) {
        [result setObject:[[[self marker] userData] valueForKey:@"uuid"] forKey:@"uuid"];
    }
    
    [[self marker] setUserData:result];
    [self replaceValue:result forKey:@"userData" notification:NO];
}

- (void)updateLocation:(id)args
{
    ENSURE_UI_THREAD(updateLocation, args);
    ENSURE_SINGLE_ARG(args, NSDictionary);
    
    id latitude = [args valueForKey:@"latitude"];
    id longitude = [args valueForKey:@"longitude"];
    id animated = [args valueForKey:@"animated"];
    id duration = [args valueForKey:@"duration"];
    id rotation = [args valueForKey:@"rotation"];
    id opacity = [args valueForKey:@"opacity"];
    
    if ([TiUtils boolValue:animated def:NO]) {
        [CATransaction begin];
        [CATransaction setAnimationDuration:[TiUtils floatValue:duration def:2000] / 1000];
        
        // Update coordinates
        if (latitude != nil && longitude != nil) {
            [[self marker] setPosition:CLLocationCoordinate2DMake([TiUtils doubleValue:latitude],[TiUtils doubleValue:longitude])];
        }
        
        // Update rotation
        if (rotation != nil) {
            [[self marker] setRotation:[TiUtils doubleValue:rotation def:0]];
        }
        
        // Update opacity
        if (opacity != nil) {
            [[self marker] setOpacity:[TiUtils floatValue:opacity def:1]];
        }
        
        [CATransaction commit];
    } else {
        [[self marker] setPosition:CLLocationCoordinate2DMake([TiUtils doubleValue:latitude],[TiUtils doubleValue:longitude])];
    }
    
    [self replaceValue:latitude forKey:@"latitude" notification:NO];
    [self replaceValue:longitude forKey:@"longitude" notification:NO];
}

- (void)setCustomView:(id)value
{
    ENSURE_UI_THREAD(setCustomView, value);
    
    id current = [self valueForUndefinedKey:@"customView"];
    [self replaceValue:value forKey:@"customView" notification:NO];
    
    if (value == [NSNull null] || value == nil) {
        [[self marker] setIconView:nil];
    } else if ([current isEqual:value] == NO) {
        [self forgetProxy:current];
        [self rememberProxy:value];
        
        [[self marker] setIconView:[(TiViewProxy *)value view]];
    }
}

@end

File: titanium-googlemaps-swift/Classes/TiGooglemapsAnnotationProxy.swift
/**
 * Axway Titanium
 * Copyright (c) 2018-present by Axway Appcelerator. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

import UIKit
import TitaniumKit
import GoogleMaps
import QuartzCore

@objc(TiGooglemapsAnnotationProxy)
public class TiGooglemapsAnnotationProxy: TiProxy {
  private var markerInternal: GMSMarker?
  private var customViewProxy: TiViewProxy?
  private var infoWindowProxy: TiViewProxy?
  
  // MARK: - Lifecycle
  
  @objc(_initWithPageContext:andMarker:)
  public func _init(withPageContext context: TiEvaluator?, andMarker marker: GMSMarker) -> Self? {
    guard let proxy = super._init(withPageContext: context) else {
      return nil
    }
    
    markerInternal = marker
    replaceValue(marker.position.latitude, forKey: "latitude", notification: false)
    replaceValue(marker.position.longitude, forKey: "longitude", notification: false)
    replaceValue(marker.title ?? NSNull(), forKey: "title", notification: false)
    replaceValue(marker.snippet ?? NSNull(), forKey: "subtitle", notification: false)
    
    ensureUserDataHasUUID()
    
    return proxy
  }
  
  // MARK: - Helpers
  
  private func runOnMainThread(wait: Bool = false, _ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      TiThreadPerformOnMainThread(block, wait)
    }
  }
  
  private func ensureUserDataHasUUID() {
    guard let marker = markerInternal else { return }
    
    if var data = marker.userData as? [String: Any] {
      if data["uuid"] == nil {
        data["uuid"] = UUID().uuidString
        marker.userData = data
      }
    } else {
      marker.userData = ["uuid": UUID().uuidString]
    }
  }
  
  private func currentCoordinate() -> CLLocationCoordinate2D {
    let latitude = TiUtils.doubleValue(value(forKey: "latitude"))
    let longitude = TiUtils.doubleValue(value(forKey: "longitude"))
    return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
  }
  
  private func updateMarkerPositionIfPossible() {
    guard let marker = markerInternal else { return }
    marker.position = currentCoordinate()
  }
  
  private func number(from value: Any?) -> NSNumber? {
    if let number = value as? NSNumber {
      return number
    }
    
    if let string = value as? String, let doubleValue = Double(string) {
      return NSNumber(value: doubleValue)
    }
    
    return nil
  }
  
  // MARK: - Public API
  
  @objc public var marker: GMSMarker {
    if let marker = markerInternal {
      return marker
    }
    
    let marker = GMSMarker()
    marker.position = currentCoordinate()
    marker.isTappable = true
    marker.userData = ["uuid": UUID().uuidString]
    markerInternal = marker
    return marker
  }
  
  @objc public func infoWindow() -> TiViewProxy? {
    return infoWindowProxy
  }
  
  @objc public func setInfoWindow(_ value: Any?) {
    infoWindowProxy = value as? TiViewProxy
    replaceValue(value, forKey: "infoWindow", notification: false)
  }
  
  @objc public func setLatitude(_ value: Any?) {
    replaceValue(value, forKey: "latitude", notification: false)
    updateMarkerPositionIfPossible()
  }
  
  @objc public func setLongitude(_ value: Any?) {
    replaceValue(value, forKey: "longitude", notification: false)
    updateMarkerPositionIfPossible()
  }
  
  @objc public func setTitle(_ value: Any?) {
    runOnMainThread {
      let title = TiUtils.stringValue(value)
      self.marker.title = title
      self.replaceValue(title ?? NSNull(), forKey: "title", notification: false)
    }
  }
  
  @objc public func setSubtitle(_ value: Any?) {
    runOnMainThread {
      let subtitle = TiUtils.stringValue(value)
      self.marker.snippet = subtitle
      self.replaceValue(subtitle ?? NSNull(), forKey: "subtitle", notification: false)
    }
  }
  
  @objc public func setZIndex(_ value: Any?) {
    runOnMainThread {
      self.marker.zIndex = Int32(TiUtils.intValue(value))
      self.replaceValue(value, forKey: "zIndex", notification: false)
    }
  }
  
  @objc public func setCenterOffset(_ value: Any?) {
    runOnMainThread {
      self.marker.infoWindowAnchor = TiUtils.pointValue(value)
      self.replaceValue(value, forKey: "centerOffset", notification: false)
    }
  }
  
  @objc public func setGroundOffset(_ value: Any?) {
    runOnMainThread {
      self.marker.groundAnchor = TiUtils.pointValue(value)
      self.replaceValue(value, forKey: "groundOffset", notification: false)
    }
  }
  
  @objc public func setImage(_ value: Any?) {
    runOnMainThread {
      self.marker.icon = TiUtils.image(value, proxy: self)
      self.replaceValue(value, forKey: "image", notification: false)
    }
  }
  
  @objc public func setCustomIcon(_ value: Any?) {
    guard let dictionary = value as? [String: Any] else { return }
    
    runOnMainThread {
      let textColor = TiUtils.colorValue(dictionary["textColor"])?.color ?? .black
      let tintColor = TiUtils.colorValue(dictionary["tintColor"])?.color
      let title = TiUtils.stringValue(dictionary["title"]) ?? ""
      let image = TiUtils.image(dictionary["image"], proxy: self)
      let font = TiUtils.fontValue(dictionary["font"])?.font() ?? UIFont.systemFont(ofSize: 12, weight: .regular)
      
      let imageView = UIImageView(image: image)
      imageView.frame = CGRect(origin: .zero, size: image?.size ?? .zero)
      if let tintColor = tintColor {
        imageView.tintColor = tintColor
        imageView.image = imageView.image?.withRenderingMode(.alwaysTemplate)
      }
      
      let labelFrame = CGRect(origin: .zero, size: CGSize(width: imageView.bounds.width, height: imageView.bounds.width))
      let titleLabel = UILabel(frame: labelFrame)
      titleLabel.font = font
      titleLabel.text = title
      titleLabel.textAlignment = .center
      titleLabel.textColor = textColor
      
      let container = UIView(frame: imageView.bounds)
      container.addSubview(imageView)
      container.addSubview(titleLabel)
      
      self.marker.iconView = container
      self.marker.tracksViewChanges = false
      self.replaceValue(dictionary, forKey: "customIcon", notification: false)
    }
  }
  
  @objc public func setPinColor(_ value: Any?) {
    runOnMainThread {
      if let color = TiUtils.colorValue(value)?.color {
        self.marker.icon = GMSMarker.markerImage(with: color)
      }
      self.replaceValue(value, forKey: "pinColor", notification: false)
    }
  }
  
  @objc public func setTouchEnabled(_ value: Any?) {
    runOnMainThread {
      self.marker.isTappable = TiUtils.boolValue(value, def: true)
      self.replaceValue(value, forKey: "touchEnabled", notification: false)
    }
  }
  
  @objc public func setFlat(_ value: Any?) {
    runOnMainThread {
      self.marker.isFlat = TiUtils.boolValue(value, def: false)
      self.replaceValue(value, forKey: "flat", notification: false)
    }
  }
  
  @objc public func setDraggable(_ value: Any?) {
    runOnMainThread {
      self.marker.isDraggable = TiUtils.boolValue(value, def: false)
      self.replaceValue(value, forKey: "draggable", notification: false)
    }
  }
  
  @objc public func setOpacity(_ value: Any?) {
    runOnMainThread {
      self.marker.opacity = Float(TiUtils.floatValue(value, def: 1))
      self.replaceValue(value, forKey: "opacity", notification: false)
    }
  }
  
  @objc public func setAnimationStyle(_ value: Any?) {
    runOnMainThread {
      let defaultAnimation = GMSMarkerAnimation.none
      let rawValue = TiUtils.intValue(value, def: Int32(defaultAnimation.rawValue))
      self.marker.appearAnimation = GMSMarkerAnimation(rawValue: UInt(rawValue)) ?? defaultAnimation
      self.replaceValue(value, forKey: "animationStyle", notification: false)
    }
  }
  
  @objc public func setRotation(_ value: Any?) {
    runOnMainThread {
      self.marker.rotation = TiUtils.doubleValue(value, def: 0)
      self.replaceValue(value, forKey: "rotation", notification: false)
    }
  }
  
  @objc public func setUserData(_ value: Any?) {
    runOnMainThread {
      var result = value as? [String: Any] ?? [:]
      if let uuid = (self.marker.userData as? [String: Any])?["uuid"] {
        result["uuid"] = uuid
      } else if result["uuid"] == nil {
        result["uuid"] = UUID().uuidString
      }
      self.marker.userData = result
      self.replaceValue(result, forKey: "userData", notification: false)
    }
  }
  
  @objc public func updateLocation(_ args: [String: Any]) {
    runOnMainThread {
      let latitude = self.number(from: args["latitude"])
      let longitude = self.number(from: args["longitude"])
      let animated = TiUtils.boolValue(args["animated"], def: false)
      let duration = TiUtils.floatValue(args["duration"], def: 2000) / 1000
      let rotation = self.number(from: args["rotation"])
      let opacity = self.number(from: args["opacity"])
      
      let updateBlock = {
        if let lat = latitude, let lon = longitude {
          let coordinate = CLLocationCoordinate2D(latitude: lat.doubleValue, longitude: lon.doubleValue)
          self.marker.position = coordinate
          self.replaceValue(lat, forKey: "latitude", notification: false)
          self.replaceValue(lon, forKey: "longitude", notification: false)
        }
        
        if let rotation = rotation {
          self.marker.rotation = rotation.doubleValue
          self.replaceValue(rotation, forKey: "rotation", notification: false)
        }
        
        if let opacity = opacity {
          self.marker.opacity = opacity.floatValue
          self.replaceValue(opacity, forKey: "opacity", notification: false)
        }
      }
      
      if animated {
        CATransaction.begin()
        CATransaction.setAnimationDuration(CFTimeInterval(duration))
        updateBlock()
        CATransaction.commit()
      } else {
        updateBlock()
      }
    }
  }
  
  @objc public func setCustomView(_ value: Any?) {
    runOnMainThread {
      let previous = self.customViewProxy
      var newProxy: TiViewProxy?
      
      if let proxy = value as? TiViewProxy {
        newProxy = proxy
        self.remember(proxy)
        self.marker.iconView = proxy.view
      } else {
        self.marker.iconView = nil
      }
      
      if let previous = previous, previous !== newProxy {
        self.forget(previous)
      }
      
      self.customViewProxy = newProxy
      self.replaceValue(value ?? NSNull(), forKey: "customView", notification: false)
    }
  }
  
  // MARK: - Overrides
  
  public override func keySequence() -> [Any]! {
    return ["latitude", "longitude"]
  }
}
